### ISSUE TITLE
loudml on OpenShift

### ISSUE THREAD

**Issue by simplelife2010** [regel/loudml#229]
Hi,

I have made some changes to the Dockerfile to deploy loudml on an OpenShift cluster. Is that something you would want in your codebase?

Bruno

**Comment by regel** [comment]
Sounds super cool. I’m happy to review and give some feedback.

**Comment by simplelife2010** [comment]
Well there is two things to take into account without overriding defaults in OpenShiift:
1. Your are not allowed to run a container as root
2. The user that runs the container is not known by name. That means he has only a random user id on the OS level.

This is what I did to make it work:
```
FROM loudml:latest-devel
RUN chgrp -R 0 /var/lib/loudml && \
    chmod -R g=u /var/lib/loudml && \
    chgrp 0 /etc/passwd && \
    chmod g=u /etc/passwd
CMD cp /etc/passwd /tmp/passwd && \
    sed -i "s/loudml:x:...:...../loudml:x:$(id -u):0/" /tmp/passwd && \
    truncate -s 0 /etc/passwd && \
    cat /tmp/passwd >> /etc/passwd && \
    rm /tmp/passwd && \
    /usr/bin/loudmld
```

This means:
1. Make /var/lib/loudml accessible for a non-privileged user:
```
chgrp -R 0 /var/lib/loudml
chmod -R g=u /var/lib/loudml
```
This is exactly what OpenShift suggests to do according to their documentation.

2. Make /etc/passwd accessible for a non-privileged user:
```
chgrp 0 /etc/passwd
chmod g=u /etc/passwd
```

3. Ensure that at runtime (when we know the user id), we hack a username into the system that is necessary for the service to run. When there is no username associated to the user id, the service will not run:
```
CMD cp /etc/passwd /tmp/passwd && \
    sed -i "s/loudml:x:...:...../loudml:x:$(id -u):0/" /tmp/passwd && \
    truncate -s 0 /etc/passwd && \
    cat /tmp/passwd >> /etc/passwd && \
    rm /tmp/passwd && \
    /usr/bin/loudmld
```

So far using this patch LoudML works well on our OpenShift cluster. What do you think?

**Comment by simplelife2010** [comment]
I admit that the modified CMD would not work on a normal Docker setup when a username exists. Perhaps we could change it so at runtime it checks whether a username exists or not.

**Comment by regel** [comment]
Hi Bruno, thanks so much for this!

Can you please provide more details about point 3, what is the error you see?  What prevents the daemon from running?

Did you look at Bitnami packaging guidelines and UID 1001? I'd like to change the Dockerfile first to enable UID 1001, and then look into OpenShift. Sounds like a plan?

**Comment by simplelife2010** [comment]
Hi,

I cannot reproduce the issue #3 any more. It  seems the service works without the modification. No idea why I had to do this in the beginning. Perhaps I had used an older loudml Docker image.

**Comment by regel** [comment]
Hi Bruno. I merged the above changes in `develop` thanks so much for lighting the way!

`/opt/vendor` contains the virtualenv of Loud ML dependencies so I used `chgrp` on this directory as well.

**Comment by regel** [comment]
Let me know if we can close this item. If you need to change the base image for OpenShift, maybe you can set "--build-arg base_image=xxxxx" during your Docker build.

**Comment by simplelife2010** [comment]
Thank you! I hope I will find some time soon to retest.

**Comment by simplelife2010** [comment]
Is the mentioned patch not yet in Dockerhub?

```
aim-loudml % docker run -it loudml/loudml:latest-devel /bin/bash
Unable to find image 'loudml/loudml:latest-devel' locally
latest-devel: Pulling from loudml/loudml
b248fa9f6d2a: Pull complete
13f9522197b3: Pull complete
f2b0222d160d: Pull complete
ea194de223a6: Pull complete
a409651de252: Pull complete
Digest: sha256:529a6de7947a167400cadd5f6bbbba421442cf777a085ba46749fac4d360ae3a
Status: Downloaded newer image for loudml/loudml:latest-devel
root@4a66d6c61ed9:/# ls -l /var/lib
total 40
drwxr-xr-x 1 root   root   4096 May  4 07:24 apt
drwxr-xr-x 2 root   root   4096 May  4 07:24 dbus
drwxr-xr-x 1 root   root   4096 May  4 07:24 dpkg
drwxrwsr-x 2 loudml loudml 4096 May  4 07:24 loudml
drwxr-xr-x 2 root   root   4096 Feb  1 17:09 misc
drwxr-xr-x 2 root   root   4096 Apr 22 00:00 pam
drwxr-xr-x 2 root   root   4096 May  4 07:23 python
drwxr-xr-x 2 root   root   4096 May  4 07:24 sgml-base
drwxr-xr-x 3 root   root   4096 Apr 22 00:00 systemd
drwxr-xr-x 2 root   root   4096 May  4 07:24 xml-core
```

**Comment by regel** [comment]
Oops. Something is odd. I should clarify this in the documentation.

The `master` branch is currently tagged latest-devel. The latest commit is from January and matches the tag you just pulled. I guess it should be renamed into 'latest' to comply with standard tag names in Docker.

Every night, `develop` is built and tagged `nightly`.

Tags available:

- loudml/loudml:nightly-noavx
- loudml/loudml:nightly-gpu
- loudml/loudml:nightly

When `develop` is merged into `master` (that's when 1.6.0 is tag is released - soon); things will be more consistent.

**Comment by simplelife2010** [comment]
Thank you, it works now.
